AppController/djinn.rb

```diff
@@ -938,19 +932,100 @@ def set_parameters(layout, options, secret)
       if PARAMETERS_AND_CLASS[key][0] == TrueClass or
         PARAMETERS_AND_CLASS[key][0] == Fixnum
         begin
-          @options[key] = val.to_s
+          newval = val.to_s
         rescue
           msg = "Warning: cannot convert '" + key + "' to string. Removing it."
           Djinn.log_warn(msg)
-          options_to_delete.push(key)
+          next
+        end
+      end
+
+      # Strings may need to be sanitized.
+      if PARAMETERS_AND_CLASS[key][0] == String
+        # Some options shouldn't be sanitize.
+        if key == 'user_commands' or key == 'azure_app_secret_key'
+          newval = val
+        # Keys have a relaxed sanitization process.
+        elsif key.include? "_key" or key.include? "EC2_SECRET_KEY"
+          newval = val.gsub(NOT_FQDN_OR_PLUS_REGEX, "")
+        else
+          newval = val.gsub(NOT_FQDN_REGEX, "")
         end
-        next
       end
+
+      Djinn.log_debug("Accepted option #{key}:#{newval}.")
+      newoptions[key] = newval
     }
-    options_to_delete.each { |key|
-      @options.delete(key)
+
+    return newoptions
+  end
+
+  def enforce_options()
+    # Set the proper log level.
+    new_level = Logger::INFO
+    new_level = Logger::DEBUG if @options['verbose'].downcase == "true"
+    @@log.level = new_level if @@log.level != new_level
+
+    # Make sure flower is running with the proper password.
+    TaskQueue.stop_flower
+    TaskQueue.start_flower(@options['flower_password']) if my_node.is_shadow?
+  end
+
+  # Validates and sets the instance variables that Djinn needs before it can
+  # begin configuring and deploying services on a given node (and if it is the
+  # first Djinn, starting up the other Djinns).
+  #
+  # Args:
+  #   layout: this is a JSON structure containing the nodes
+  #     informations (IPs, roles, instance ID etc...). These are the nodes
+  #     specified in the AppScalefile at startup time.
+  #   options: this is a Hash containing all the options and credentials
+  #     (for autoscaling) pertinent to this deployment.
+  def set_parameters(layout, options, secret)
+    return BAD_SECRET_MSG unless valid_secret?(secret)
+
+    # options is a JSON string that will be loaded into a Hash.
+    if options.class != String
+      msg = "Error: options wasn't a String, but was a " +
+            options.class.to_s
+      Djinn.log_error(msg)
+      return msg
+    end
+    begin
+      opts = JSON.load(options)
+    rescue JSON::ParserError
+      msg = "Error: got exception parsing JSON options."
+      Djinn.log_error(msg)
+      return msg
+    end
+    if opts.nil? || opts.empty?
+      Djinn.log_warn("Empty options: using defaults (autoscale is not enabled).")
+    elsif opts.class != Hash
+      msg = "Error: options is not a Hash."
+      Djinn.log_error(msg)
+      return msg
+    else
+      @options = check_options(opts)
+    end
+
+    # Let's validate we have the needed options defined.
+    ['keyname', 'login', 'table'].each { |key|
```

Comment:
we do, but I think is still valuable to leave it for community member wanting to work on a different database. Moreover changing it is a bit more complicated at this point.